text/utilities/0000-chmod.md

```diff
@@ -0,0 +1,81 @@
+# Chmod Buildpack
+
+## Summary
+
+A chmod buildpack that allows users to change directories and files (produced at buildtime) UNIX permissions.
+
+This buildpack would be configured to run last, so that it would be able to change any directories and files produced by previous buildpacks.
+
+## Motivation
+
+Buildpacks and Builders rely on Stacks.
+
+Stacks define build time and runtime base images.
+
+Some stacks do not use the same user / UID between build and run images, [for security reasons](https://github.com/paketo-buildpacks/rfcs/blob/main/text/0045-user-ids.md), since a runtime user should just be able to run a process; not create directories nor install packages.
+
+Unfortunately, some applications, like `dist-zip` archives, do need to create directories and files: they run setup scripts during runtime that fail because they can't write folders nor files, since they were run with a runtime user, who can be restricted with some Stacks. 
+
+Similarly, some scripts that need to be run at runtime are only executable by the build time user, and would need to be executable by the runtime user.
+
+## Detailed Explanation
+
+N/A
+
+## Rationale and Alternatives
+
+We have considered solving the issue at different levels:
+
+* At the stack level: reverting to former stacks that define the same user at build time and runtime.
+
+After all, previous stacks like Ubuntu Bionic did not differentiate the runtime user from the build time user, both being named `cnb` with the UID 1000
+
+We began noticing this issue with the latter stacks (some based on Ubuntu Jammy) that guarantee [a more secure environment](https://github.com/buildpacks/rfcs/blob/main/text/0085-run-uid.md) by having the build time user set to UID 1002, and runtime user set to 1000 at runtime.
+
+* At the buildpack level: [implementing special flags in some buildpacks](https://github.com/paketo-buildpacks/dist-zip/pull/174) that would allow users of those buildpacks to make some folders group writable
+
+But then, we discovered that it was not enough, some other flag would need to be implemented to let the build time user run a different set of scripts; and of course, we would need to replicate this behavior in several other buildpacks packaging applications that set themselves up at runtime.
+
+Allowing the user to keep on using the same secure Stacks, and unmodified buildpacks, using a chmod buildpack at the end of their build seems like the best solution
+
+## Implementation
+
+The chmod buildpack would be added last to builders and would be activated if a specific flag is set.
```

Comment:
There was some strong feedback in the working group meeting about ensuring it is hard/impossible for users to accidentally execute this buildpack with dangerous permissions (e.g. `chmod -R 755 /workspace`).

So, it sounds like it would be prudent to keep this buildpack out of the builders, and require users to opt-in by explicitly providing the buildpack in a custom build order.

Reply:
cc @BarDweller @jpena-r7

Reply:
I really don't like this buildpack.. I mean, I totally get why it's needed.. but the potential for damage here is huge. 

My main concern is that with this being a standard offering, that you'd see a propagation of 'just chmod everything 755' style 'fixes' starting to be applied by users who discover this 'makes it work' without understanding what they've done. 

Unix permissions are not simple, I get that, but this isn't a fix for the permissions being bad, this is a hammer that will be applied indiscriminately and by people whose main concern is 'making it all work', potentially with little to no understanding of what was broken, or the security, or application implications to the broad permission changes they just applied. (You can consider that this dumps back on the user the responsibility to understand the container being built, which they already made the choice to delegate to buildpacks)

In an ideal world, we shouldn't need this at all. The relevant buildpacks should be able to determine the appropriate permissions for their files to ensure that the application is able to function at runtime. If this is not possible today, then we really need to start identifying those scenarios, and working out how to deal with them. 

If we consider that this is needed because of certain scenarios, AND we envisage those scenarios may be fixed in future, then I'd argue this buildpack should probably offer 'presets' that are known to fix permissions.. so that in future, as individual scenarios are fixed upstream, or no longer a concern, that this buildpack is able to no-op the changes it would make. 

When individual files/dirs are altered via this buildpack, without a connection to why, it'll be pretty tough in future for an end user to know what to remove, and when to remove, from this buildpacks config if an underlying issue is resolved by a framework/runtime.

That said.. I do see that this is still needed.. today.. because this is the simplest/easiest option to fix the permissions issues we see today. The thought that this buildpack can very easily be unintentionally used to undo all the careful preparation made by prior buildpacks, and potentially expose the application to content it should never have been able to read or serve is pretty chilling though. 

Re disabling at the platform level:
I don't think disabling it at a platform level makes sense.. you either need the changes to make the application work, or you don't. If a platform chose to disable this buildpack, it would just become a less useful platform.. sure, maybe it's more secure, but if its also no longer able to build a working application, wouldn't someone just switch to something else until it works? 

Re tracking actions via a label
I don't like the idea of unbounded labels either.. just setting something with a summary of how many files/dirs were altered, might give an indication of how many things were affected.. Also the build log should explicitly log out the files/dirs altered saying from what, to what permissions. This will at least allow the log to be used for diagnosis of how files ended up with unexpected permissions in the future (Imagine you are a buildpack author handling an Issue raised that an app built with your buildpack isn't functioning correctly, how would you know this buildpack had potentially altered the permissions of the files you created.. etc)


Overall, the main thing here is that the documentation for this should be explicitly clear regarding the potential risks of using this buildpack. This should include the RFC itself containing a clear warning about the potential consequences of applying permissions recursively. 

Understood that this is just a tool, and it's how you use it that makes it dangerous, not the tool itself..  However, I've seen too many 'fix it by `chmod -R 755 /directory` ' hack-solutions that end up causing disaster for production containers.

Reply:
>My main concern is that with this being a standard offering, that you'd see a propagation of 'just chmod everything 755' style 'fixes' starting to be applied by users who discover this 'makes it work' without understanding what they've done.

I disagree. When things generally work out of the box, you don't need this and won't use this. Permissions are not an issue *most* of the time with buildpacks, so most of the time this won't even be used. No one can see the future, but that's my theory.

>In an ideal world, we shouldn't need this at all. The relevant buildpacks should be able to determine the appropriate permissions for their files to ensure that the application is able to function at runtime. If this is not possible today, then we really need to start identifying those scenarios, and working out how to deal with them.

This can't happen 100% of the time. There are edge cases where the buildpack cannot know the permissions to apply. If a user ships a script they want to shell out and run, we can't mark that executable for them.

The same happens with the dist-zip buildpack. That format uses scripts to execute the application and there are some cases where only the user really knows how the permissions should be set (it's further complicated by using a zip format that doesn't persist permissions).

> The thought that this buildpack can very easily be unintentionally used to undo all the careful preparation made by prior buildpacks, and potentially expose the application to content it should never have been able to read or serve is pretty chilling though.

I think this is a bit sensational. Buildpacks make an effort so that things just work and are secure out of the box, but there are still plenty of ways for a dev to mess things up. Devs can include inline buildpacks, or just make their own buildpacks. Most build tools also offer a way to run arbitrary commands at build time (Maven, Gradle, npm, composer, etc...).

The nice thing about this approach is that we can offer some controls, like logging what get's altered and when things get altered, as well as affixing labels to the generated images so that it's clear when something's been altered. That's extremely useful as a platform operator because you can now audit when someone's doing this or even block it by prohibiting containers with those labels from being run. That's way better than the alternative approaches users would have to take without this buildpack, and way better than the standard Dockerfile approach.

> Re disabling at the platform level:

The point is that a platform operator can make this choice if they want to prevent devs from doing this. It's additional flexibility and control, which is something that buildpacks offer above and beyond Dockerfiles.

> Overall, the main thing here is that the documentation for this should be explicitly clear regarding the potential risks of using this buildpack. This should include the RFC itself containing a clear warning about the potential consequences of applying permissions recursively.

I don't mind adding warnings to the documentation. It's important to let users know when to be wary but we shouldn't put anything so over the top as to scare users off. We should also be sure to document the labels/audit logging/etc.. so that they can see how this option can effectively be used.

Reply:
Added big warnings as a first chapter.